vms/evm/imports_test.go

```diff
@@ -0,0 +1,168 @@
+// Copyright (C) 2019-2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package tests
+
+import (
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"os"
+	"path/filepath"
+	"regexp"
+	"slices"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/ava-labs/avalanchego/utils/set"
+)
+
+// TestDoNotImportFromGraft ensures proper import rules for graft packages:
+// - graft/coreth can be imported anywhere EXCEPT vms/evm (but vms/evm/emulate is an exception)
+// - graft/subnet-evm cannot be imported anywhere EXCEPT vms/evm/emulate
+func TestEnforceGraftImportBoundaries(t *testing.T) {
+	graftRegex := regexp.MustCompile(`^github\.com/ava-labs/avalanchego/graft(/|$)`)
+
+	// Find all graft imports in the entire repository
+	foundImports, err := findImportsMatchingPattern("..", graftRegex, func(path string, importPath string, _ *ast.ImportSpec) bool {
+		// Skip generated files and test-specific files
+		filename := filepath.Base(path)
+		if strings.HasPrefix(filename, "gen_") ||
+			strings.Contains(path, "graft/*/core/main_test.go") ||
+			strings.Contains(path, "graft/*/tempextrastest/") {
+			return true
+		}
+
+		// Skip files in the graft directory itself - they can import from graft
+		if strings.Contains(path, "/graft/") {
+			return true
+		}
+
+		isInEmulate := strings.Contains(path, "/vms/evm/emulate/")
+		isInVmsEvm := strings.Contains(path, "/vms/evm/")
+		isCoreth := strings.Contains(importPath, "/graft/coreth")
+		isSubnetEVM := strings.Contains(importPath, "/graft/subnet-evm")
+
+		// graft/coreth: illegal in vms/evm except vms/evm/emulate
+		if isCoreth && isInVmsEvm && !isInEmulate {
+			return false
+		}
+
+		// graft/subnet-evm: illegal everywhere except vms/evm/emulate
+		if isSubnetEVM && !isInEmulate {
+			return false
+		}
+
+		return true
+	})
+	require.NoError(t, err, "Failed to find graft imports")
+
+	if len(foundImports) == 0 {
+		return // no violations found
+	}
+
+	header := "Graft import rules violated!\n" +
+		"Rules:\n" +
+		"  - graft/coreth can be imported anywhere EXCEPT vms/evm (but vms/evm/emulate is an exception)\n" +
+		"  - graft/subnet-evm cannot be imported anywhere EXCEPT vms/evm/emulate\n\n"
+	require.Fail(t, formatImportViolations(foundImports, header))
+}
+
+// TestDoNotImportLibevmPseudo ensures that no code in the repository imports
+// the libevm/pseudo package, which is for internal libevm use only.
+func TestDoNotImportLibevmPseudo(t *testing.T) {
```

Comment:
```suggestion
func TestDoNotImportLibEVMPseudo(t *testing.T) {
```

Although the project name is stylised `libevm`, initialisms in Go identifiers should have all initials with the same capitalisation.

Reply:
Outdated